Topics/Tech_Stacks/Web-Sockets.md

```diff
@@ -0,0 +1,162 @@
+
+## Understanding WebSockets
+WebSockets are a powerful protocol embedded within the familiar web technology stack, enabling a significant leap in communication capabilities over the Internet. They provide a way to overcome the limitations of the traditional HTTP request-response model, especially when it comes to real-time, low-latency communication.
+
+Unlike HTTP, where a new connection is established for each request and response, WebSockets establish a single, persistent connection between the client and the server. This connection is initiated by the client sending a special type of HTTP request, asking the server to upgrade the connection to a WebSocket connection.
+
+### How the WebSocket Connection is Established:
+1. **Client Handshake Request:** The client initiates the connection by sending a WebSocket handshake request to the server. This request looks like a standard HTTP request but includes an `Upgrade: websocket` header that signals the desire to establish a WebSocket connection.
+   
+2. **Server Handshake Response:** If the server supports WebSockets and accepts the connection request, it responds with a handshake of its own, confirming the upgrade from HTTP to WebSockets.
+   
+3. **Persistent Connection:** With the handshake complete, the connection remains open, and data can now be sent in either direction at any time, allowing for two-way, **real-time communication**.
+
+### WebSocket Communication:
+Once established, the WebSocket connection allows messages to be passed back and forth between the client and server with minimal overhead. Each message is framed in a standardized format, which can be quickly parsed and processed by both parties. This efficient messaging system is key to the protocol's ability to support high-frequency, real-time communication scenarios.
+
+WebSockets are particularly adept at handling use cases where the server needs to send data to the client proactively, without the client first making a request. This capability has led to the widespread adoption of WebSockets in various applications that require real-time functionality, such as live notifications, instant messaging, and collaborative editing platforms.
+
+In the context of CSC301, understanding WebSockets is essential for building modern applications where real-time user interaction is a core requirement. With WebSockets, software engineers can create responsive and interactive user experiences that were once difficult to achieve with traditional web technologies.
```

Comment:
Can you give an example for this? Just one line example could be helpful

Reply:
Do you mean example use cases?

Reply:
Hi, sorry for the miscommunication. I must have misread the content. You have clearly given an example of WebSockets "live notification"